# Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# Import Dataframes

In [2]:
# define path
path=(r'C:\Users\Admin\Desktop\Data Tasks\Data Immersion\Achievement 4\Instacart Basket Analysis\02 data')

In [3]:
# importing products.csv
df_prods=pd.read_csv(os.path.join(path, 'Original Data','products.csv'),index_col=False)

In [4]:
# importing orders.csv
df_ords=pd.read_csv(os.path.join(path,'Prepared Data','orders_wrangled.csv'),index_col=0)

# Direction 1:

#### If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.

## Mixed-type Data

In [5]:
# checking the columns for mixed-type data
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)


product_name


##### note: we'll be handling dtype of product_name matter in the topic below: 

## Missing Values

In [6]:
# finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [7]:
#displaying missing values by creating a subset
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [8]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

##### note: the mixed-type data in the product_name depicted above were the rows with NaN values

In [9]:
#creating a new df for products excluding the NaN values
df_prods_clean=df_prods[df_prods['product_name'].isnull()==False]

In [10]:
#checking row count before removing NaN
df_prods.shape

(49693, 5)

In [11]:
#after removing Nan
df_prods_clean.shape

(49677, 5)

## Duplicates

In [12]:
#finding duplicated rows
df_dups=df_prods_clean[df_prods_clean.duplicated()]

In [13]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [14]:
#dropping the duplicates
df_prods_clean_no_dups=df_prods_clean.drop_duplicates()

In [15]:
#checking the row count before dropping dupes
df_prods_clean.shape

(49677, 5)

In [16]:
# after dropiing dupes
df_prods_clean_no_dups.shape

(49672, 5)

# Direction 2:

#### Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
Tip: Keep an eye on min and max values!

In [17]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#### there is missing data in column days_since_prior_order since the count row is less for this column

In [18]:
#checking the NAN for days_since_prior_order
df_ords_nan=df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [19]:
df_ords_nan.shape

(206209, 6)

#### there are total of 206209 rows with missing data of the time since last order. But the missing data is not crucial for our analysis so I will make no changes for this column or rows. There are no issues with the max and min in df_ords. I will check max and min for dataframe products

In [20]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### max 99999 in the prices column appear to be too high.

In [21]:
#checking the product with the price 99999
df_max_price=df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices']==99999]

In [22]:
df_max_price

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

In [23]:
df_max_price.prices=pd.NA

C:\Users\Admin\AppData\Local\Temp\ipykernel_3124\867930473.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_max_price.prices=pd.NA


#### most probably the 99999 stands for the uknown value, in order to not have it involved in analytical calculations it would be better to change the value to NaN or to string

In [24]:
df_max_price

product_id           product_name  aisle_id  department_id prices
33666       33664  2 % Reduced Fat  Milk        84             16   <NA>

In [25]:
df_max_price.dtypes

product_id        int64
product_name     object
aisle_id          int64
department_id     int64
prices           object
dtype: object

# Direction 3:

#### Check for mixed-type data in your df_ords dataframe

In [26]:
#checking for mixed-type data
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)


# Direction 4:

#### If you find mixed-type data, fix it. The column in question should contain observations of a single data type

In [27]:
df_ords.dtypes

order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

#### result: all the columns are consistent, no mixed-type data found

# Direction 5:

#### Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [28]:
#checking for null value
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [29]:
df_ords.shape

(3421083, 6)

# Direction 6:

#### Address the missing values using an appropriate method.
In a markdown cell, explain why you used your method of choice.

#### Once again we see here the missing values for days_since_prior_order. While it makes up to 6% of the total number of rows, leaving data as it is is better than tring to delete or fix them. If we have to cunduct an analysis specifically on the days since prior order, we might take into consideration using the mean or median as a replacement for the missing values.

# Direction 7:

#### Run a check for duplicate values in your df_ords data.
In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [30]:
#checking for duplicates
df_ords_dups=df_ords[df_ords.duplicated()]

In [31]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

# Direction 8:

#### Address the duplicates using an appropriate method.
In a markdown cell, explain why you used your method of choice.

#### Result: no duplicates were found in orders dataframe

# Direction 9:

#### Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [32]:
#exporting df_prods
df_prods_clean_no_dups.to_csv(os.path.join(path,'Prepared Data','df_prods_clean.csv'))

In [33]:
#exporting df_ords
df_ords.to_csv(os.path.join(path,'Prepared Data','df_ords_clean.csv'))

##### df_ords consistency check row count

In [34]:
df_ords.shape

(3421083, 6)

##### df_prods consistency check row count

In [35]:
df_prods_clean_no_dups.shape

(49672, 5)

In [36]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0